### PR TITLE
Improve performance of project file watch setup

### PR DESCRIPTION
**Pull request by rtpg**
This improves performance through a couple of strategies:

 - Only check if a directory is an ignored directory once. Previously  the filter was in multiple parts of the the chain.
 - When setting up watches, only generate a list of directories (as files end up getting filtered out later anyways)
 - Directly list directory contents when recursively fetching items,  instead of relying on completion code

(See related discussion [here](https://github.com/emacs-lsp/lsp-mode/issues/2741#issuecomment-808828943))

### CODE REVIEW

**Comment by rtpg** [comment]
I'm having a hard time understanding these parse failures, looks like _something_ is wrong with Emacs 26.x in my thing but I can't quite figure out what

File: lsp-mode.el [hunk 1]
```diff
@@ -1725,41 +1727,32 @@ regex in IGNORED-FILES."
...
     (lsp-log "Creating watch for %s" dir)
     (when (or
            (not warn-big-repo?)
            (not lsp-file-watch-threshold)
-           (let ((number-of-files (length (lsp--directory-files-recursively dir ".*" ignored-directories t))))
+           (let ((number-of-files (length dirs-to-watch)))
              (or
               (< number-of-files lsp-file-watch-threshold)
               (condition-case _err
                   (lsp--ask-about-watching-big-repo number-of-files dir)
```
Comment (yyoncho) [inline lsp-mode.el:1740]: please fix the wording in that method because we are no longer counting files.

File: lsp-mode.el [hunk 2]
```diff
@@ -1713,6 +1686,35 @@ Do you want to watch all files in %s? "
...
+
+(defun lsp--path-is-watchable-directory (path dir ignored-directories)
+  "Figure out whether PATH (inside of DIR) is meant to have a file watcher set.
+IGNORED-DIRECTORIES is a list of regexes to filter out directories we don't want to watch."
+  (let
+      ((full-path (f-join dir path)))
+    (and (f-dir-p full-path)
+         (not (equal path "."))
+         (not (equal path ".."))
+         (not (lsp--string-match-any ignored-directories full-path))
```
Comment (yyoncho) [inline lsp-mode.el:1698]: always close the brackets on the same line, here and on a few more places in the PR.

**Review by yyoncho — COMMENTED** [review]
Looks good. I am still to perform manual testing.

@jcs090218 any idea why CI is red? Seems like we are fetching a package that depends on emacs 27.1 and all 26.x are failing?

**Comment by rtpg** [comment]
Sorry, forgot to comment but I went over the comments and tried to clean up the indentation and the naming (also updated the file warning to try and make it clearer that this is a directory count)

**Review by yyoncho — APPROVED** [review]
Thank you for contributing to `lsp-mode` and sorry for the delay.

**Pull request merged**